### DISCUSSION TITLE
Trame 2 and pushing updates to clients

### DISCUSSION THREAD

**Discussion by Pasarus** [Kitware/trame#69] [Q&A] [answered]
Hi, I have developed an application using trame v1 mostly 1.19.4 but functional and untested in 1.20. 

I am looking into migrating to v2 due to the addition of matplotlib support and it already being present in the project for other applications, so it makes sense to use this instead of plotly from v1 right? My users are more likely to have used and seen matplotlib also.

I have a few issues with the migration though, specifically, I have become dependent on a function that will update the client's view of the drawer layout as it changes based on the current context/state of the viewer and selected options, for example, I want to disable some of the UI elements if a specific VSwitch is set to False, as it turns off/hides the vtkActor they are relevant to in the VTK viewer.

To achieve disabled objects I am at present in Trame v1 recreating the layout and "pushing" it to the client using update_layout, from the main trame module, i.e `from trame import update_layout`. Is it possible to still use this behaviour with a different function? if not can it be added into Trame v2?

Thank you! I am hoping I just missed something obvious.

**Comment by jourdain** [comment, marked as answer]
I'm not sure if there is much change between v1 to v2 regarding plotly vs matplotlib. But pick the library that make sense to you regardless of trame version. I personally prefer plotly as it provide more capabilities in term of interactions (selection...).

To dynamically update the layout with trame v2 you have two options. 

1. Using the layout context manager to automatically flush its content once done with changes...

```
with layout:
   with ...widget...:
      pass # do the change
```

2. Calling `layout.flush_content()`

HTH,

Seb
